test/e2e/storage/testsuites/volumelimits.go

```diff
@@ -0,0 +1,335 @@
+			if err != nil {
+				cleanupErrors = append(cleanupErrors, fmt.Sprintf("failed to delete pod %s: %s", l.runningPod.Name, err))
+			}
+			err = l.cs.CoreV1().Pods(l.ns.Name).Delete(l.unschedulablePod.Name, nil)
+			if err != nil {
+				cleanupErrors = append(cleanupErrors, fmt.Sprintf("failed to delete pod %s: %s", l.unschedulablePod.Name, err))
+			}
+			framework.ExpectNoError(err, "failed cleaning up the unschedulable pod")
+			for _, pvc := range l.pvcs {
+				err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Delete(pvc.Name, nil)
+				if err != nil {
+					cleanupErrors = append(cleanupErrors, fmt.Sprintf("failed to delete PVC %s: %s", pvc.Name, err))
+				}
+			}
+
+			// Wait for the PVs to be deleted. It includes also pod and PVC deletion because of PVC protection.
+			// We use PVs to make sure that the test does not leave orphan PVs when a CSI driver is destroyed
+			// just after the test ends.
+			err = wait.Poll(5*time.Second, testSlowMultiplier*framework.PVDeletingTimeout, func() (bool, error) {
+				existing := 0
+				for _, pvName := range l.pvNames.List() {
+					_, err = l.cs.CoreV1().PersistentVolumes().Get(pvName, metav1.GetOptions{})
+					if err == nil {
+						existing++
+					} else {
+						if errors.IsNotFound(err) {
+							l.pvNames.Delete(pvName)
+						} else {
+							e2elog.Logf("Failed to get PV %s: %s", pvName, err)
+						}
+					}
+				}
+				if existing > 0 {
+					e2elog.Logf("Waiting for %d PVs to be deleted", existing)
+					return false, nil
+				}
+				return true, nil
+			})
+			if err != nil {
+				cleanupErrors = append(cleanupErrors, fmt.Sprintf("timed out waiting for PVs to be deleted: %s", err))
+			}
+			if len(cleanupErrors) != 0 {
+				framework.ExpectNoError(fmt.Errorf("test cleanup failed: " + strings.Join(cleanupErrors, "; ")))
+			}
+		}()
+
+		// Create <limit> PVCs for one gigantic pod.
+		for i := 0; i < limit; i++ {
+			ginkgo.By(fmt.Sprintf("Creating volume %d/%d", i+1, limit))
+			pvc := framework.MakePersistentVolumeClaim(framework.PersistentVolumeClaimConfig{
+				ClaimSize:        dDriver.GetClaimSize(),
+				StorageClassName: &l.resource.sc.Name,
+			}, l.ns.Name)
+			pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(pvc)
+			framework.ExpectNoError(err)
+			l.pvcs = append(l.pvcs, pvc)
+		}
+
+		// Create the gigantic pod with all <limit> PVCs
+		ginkgo.By("Creating pod")
+		pod := framework.MakeSecPod(l.ns.Name, l.pvcs, nil, false, "", false, false, framework.SELinuxLabel, nil)
+		// Use affinity to schedule everything on the right node
+		selection := framework.NodeSelection{}
+		framework.SetAffinity(&selection, nodeName)
+		pod.Spec.Affinity = selection.Affinity
+		l.runningPod, err = l.cs.CoreV1().Pods(l.ns.Name).Create(pod)
+		framework.ExpectNoError(err)
+
+		l.pvNames = sets.NewString()
+		ginkgo.By("Waiting for all PVCs to get Bound")
+		err = wait.Poll(5*time.Second, testSlowMultiplier*framework.PVBindingTimeout, func() (bool, error) {
```

Comment:
is there an existing helper function for this?

Reply:
Not for array of PVCs, only single PVC.

Reply:
+ I need to collect PV names to make sure they're deleted before the test ends. It takes ages and the test suite could end before the PVs are detached + deleted.

Reply:
Moved to separate func.